### PR TITLE
Add Block::shrinkToFit() method and Call it in HashJoin Lazily

### PR DESCRIPTION
**Pull request by xapq**
### Changelog category:
- Performance Improvement

### Changelog entry:
Added IColumn::shrinktoFit method, which is now used unconditionally in HashJoin.

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
This is an automated comment for commit 185f8241e19363598ed45d0c9d6dcd03697be650 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50112/185f8241e19363598ed45d0c9d6dcd03697be650/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
</table>

File: src/Columns/IColumn.h [hunk 1]
```diff
@@ -397,6 +397,16 @@ class IColumn : public COW<IColumn>
     /// It affects performance only (not correctness).
     virtual void reserve(size_t /*n*/) {}
 
+    /// Requests the removal of unused capacity.
+    /// It is a non-binding request to reduce the capacity of the underlying container to its size.
+    virtual void shrinkToFit()
+    {
+      /// Create a new column with the same values and move it into the initial column
+      auto temporary_copy = cloneEmpty();
+      temporary_copy->insertRangeFrom(*this, 0, size());
```
Comment (alexey-milovidov) [inline src/Columns/IColumn.h:406]: This looks way too inefficient. Check if it's worth adding specializations in various column types.

File: src/Columns/IColumn.h [hunk 2]
```diff
@@ -397,6 +397,16 @@ class IColumn : public COW<IColumn>
     /// It affects performance only (not correctness).
     virtual void reserve(size_t /*n*/) {}
 
+    /// Requests the removal of unused capacity.
+    /// It is a non-binding request to reduce the capacity of the underlying container to its size.
+    virtual void shrinkToFit()
```
Comment (alexey-milovidov) [inline src/Columns/IColumn.h:402]: Why a mutable interface?

File: src/Core/Block.h [hunk 3]
```diff
@@ -151,6 +151,17 @@ class Block
     /** Get a block with columns that have been rearranged in the order of their names. */
     Block sortColumns() const;
 
+    /** See IColumn::shrinkToFit() */
+    void shrinkToFit()
+    {
+        for (auto & elem : data)
+        {
+            auto mutate_column = IColumn::mutate(std::move(elem.column));
```
Comment (alexey-milovidov) [inline src/Core/Block.h:159]: .

**Comment by Yur3k** [comment]
Memory usage decreased almost two-fold. Tested using [web analytics dataset](https://datasets.clickhouse.com/hits/tsv/hits_v1.tsv.xz)

yurchik.ru-central1.internal :) SELECT query, memory_usage, query_duration_ms FROM system.query_log WHERE type = 'QueryFinish' ORDER BY query_start_time DESC LIMIT 7 FORMAT Vertical;

SELECT
    query,
    memory_usage,
    query_duration_ms
FROM system.query_log
WHERE type = 'QueryFinish'
ORDER BY query_start_time DESC
LIMIT 7
FORMAT Vertical

Query id: 7bcf5cd6-8b3e-4414-86c2-0ecf8ee5fb4d

Row 1:
──────
query:             SELECT
        UserID,
        h1.Referer,
        h1.URL,
        h2.URL
FROM test.hits AS h1
INNER JOIN test.hits AS h2 ON h1.UserID = h2.UserID AND h1.URL = h2.Referer
WHERE h1.URL != '' AND h2.URL != '' AND h1.Referer != '' AND h2.Referer != ''
ORDER BY UserID ASC
LIMIT 1
memory_usage:      2475257308
query_duration_ms: 2747

...

Row 4:
──────
query:             set use_shrink_to_fit_in_hash_join=0
memory_usage:      0
query_duration_ms: 0

Row 5:
──────
query:             SELECT
        UserID,
        h1.Referer,
        h1.URL,
        h2.URL
FROM test.hits AS h1
INNER JOIN test.hits AS h2 ON h1.UserID = h2.UserID AND h1.URL = h2.Referer
WHERE h1.URL != '' AND h2.URL != '' AND h1.Referer != '' AND h2.Referer != ''
ORDER BY UserID ASC
LIMIT 1
memory_usage:      1328365086
query_duration_ms: 2877

...

7 rows in set. Elapsed: 0.004 sec.

File: src/Columns/IColumn.h (hunk 2 above)
Comment (Yur3k) [reply src/Columns/IColumn.h]: Changed to immutable

File: src/Columns/IColumn.h (hunk 1 above)
Comment (Yur3k) [reply src/Columns/IColumn.h]: Optimized

**Comment by vdimir** [comment]
Overall looks good. I'm still a bit concerned about possible performance impact.

I compared result on synthetic test:


```sql
CREATE TABLE t1 (k UInt64, val String) ENGINE = MergeTree ORDER BY k;
INSERT INTO t1 SELECT
    sipHash64(number, 1) % 10_000_000 as k,
    randomPrintableASCII(100) as val
FROM numbers_mt(10_000_000);

CREATE TABLE t2 (k UInt64, val String) ENGINE = MergeTree ORDER BY k;
INSERT INTO t2 SELECT
    sipHash64(number, 2) % 10_000_000 as k,
    randomPrintableASCII(100) as val
FROM numbers_mt(10_000_000);
```

Run `clickhouse-benchmark`:

```bash
$ ~/clickhouse benchmark -q "SELECT 0, sum(ignore(*)) FROM t1 JOIN t2 USING (k)" --port=9000
```

Current version:

```
localhost:9000, queries 22, QPS: 0.556, RPS: 11118250.434, MiB/s: 1240.573, result RPS: 0.556, result MiB/s: 0.000.

0.000%		1.701 sec.
10.000%		1.732 sec.
20.000%		1.740 sec.
30.000%		1.740 sec.
40.000%		1.745 sec.
50.000%		1.754 sec.
60.000%		1.768 sec.
70.000%		1.771 sec.
80.000%		1.777 sec.
90.000%		1.870 sec.
95.000%		2.044 sec.
99.000%		2.350 sec.
99.900%		2.350 sec.
99.990%		2.350 sec.

---

SELECT formatReadableSize(min(memory_usage)), formatReadableSize(max(memory_usage)), count()
FROM system.query_log
WHERE (query = 'SELECT 0, sum(ignore(*)) FROM t1 JOIN t2 USING (k)') AND (type = 'QueryFinish')

┌─formatReadableSize(min(memory_usage))─┬─formatReadableSize(max(memory_usage))─┬─count()─┐
│ 2.34 GiB                              │ 2.35 GiB                              │      22 │
└───────────────────────────────────────┴───────────────────────────────────────┴─────────┘

```


Patch:

```
localhost:9001, queries 22, QPS: 0.448, RPS: 8960110.400, MiB/s: 999.768, result RPS: 0.448, result MiB/s: 0.000.

0.000%		2.043 sec.
10.000%		2.064 sec.
20.000%		2.080 sec.
30.000%		2.139 sec.
40.000%		2.150 sec.
50.000%		2.168 sec.
60.000%		2.216 sec.
70.000%		2.262 sec.
80.000%		2.290 sec.
90.000%		2.359 sec.
95.000%		2.683 sec.
99.000%		2.949 sec.
99.900%		2.949 sec.
99.990%		2.949 sec.

---

SELECT formatReadableSize(min(memory_usage)), formatReadableSize(max(memory_usage)), count()
FROM system.query_log_1
WHERE (query = 'SELECT 1, sum(ignore(*)) FROM t1 JOIN t2 USING (k)') AND (type = 'QueryFinish')

┌─formatReadableSize(min(memory_usage))─┬─formatReadableSize(max(memory_usage))─┬─count()─┐
│ 2.02 GiB                              │ 2.02 GiB                              │      22 │
└───────────────────────────────────────┴───────────────────────────────────────┴─────────┘
```

So, seems slowdown can be up to 20% (probably depends on how much columns we select and how big they are, copying of huge strings may be expensive) in exchange for reduced memory usage (15% for this case).

But new setting will be too specific and probably user don't want to tune such things. Maybe we can come up with some heuristic, estimate amount of memore before shrinking (comparing allocating vs consumed memory for block) or depend on `max_bytes_in_join` to decide then to start shrinking (for example lazily shrink all stored blocks we reach the limit 90%)

**Comment by alexey-milovidov** [comment]
@vdimir I like the idea of lazily shrinking. Let's do something simple: - the current tracked memory usage is more than half of the maximum - shrink.

**Comment by alexey-milovidov** [comment]
We forgot to close it.
Merged in https://github.com/ClickHouse/ClickHouse/pull/54584

**Pull request closed without merging**